include/deal.II/optimization/line_minimization.h

```diff
@@ -0,0 +1,501 @@
+   * ($f(x\_low)<f(x\_hi)$) and the gradient $g(x\_low)$, return the local
+   * minimum of the quadratic interpolation function.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  quadratic_fit(const NumberType x_low,
+                const NumberType f_low,
+                const NumberType g_low,
+                const NumberType x_hi,
+                const NumberType f_hi);
+
+  /**
+   * Given $x\_low$ and $x\_hi$ together with values of function
+   * ($f(x\_low)<f(x\_hi)$) and its gradients ($g(x\_low)*g(x\_hi) < 0$) at
+   * those points return the local minimum of the cubic interpolation function.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  cubic_fit(const NumberType x_low,
+            const NumberType f_low,
+            const NumberType g_low,
+            const NumberType x_hi,
+            const NumberType f_hi,
+            const NumberType g_hi);
+
+  /**
+   * Find the minimizer of a cubic polynomial that goes through the
+   * points $f\_low=f(x\_low)$, $f\_hi=f(x\_hi)$ and $f\_rec(x\_rec)$
+   * and has derivatve $g\_low$ at $x\_low$.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  cubic_fit_three_points(const NumberType x_low,
+                         const NumberType f_low,
+                         const NumberType g_low,
+                         const NumberType x_hi,
+                         const NumberType f_hi,
+                         const NumberType x_rec,
+                         const NumberType f_rec);
+
+  /**
+   * Perform cubic fit if gradients have different sign, otherwise fallback
+   * to quadratic.
+   */
+  template <typename NumberType>
+  NumberType
+  poly_fit(const NumberType                        x_low,
+           const NumberType                        f_low,
+           const NumberType                        g_low,
+           const NumberType                        x_hi,
+           const NumberType                        f_hi,
+           const NumberType                        g_hi,
+           const NumberType                        x_rec,
+           const NumberType                        f_rec,
+           const unsigned int                      i,
+           const std::pair<NumberType, NumberType> bounds);
+
+  /**
+   * Perform a line search in $(0,max]$ with strong Wolfe conditions
+   * \f[
+   * f(\alpha) \le f(0) + \alpha \mu f'(0) \\
+   * |f'(\alpha)| \le \eta |f'(0)|
+   * \f]
+   * using one dimension
+   * functions @p func and a function @p interpolate to choose a new point
+   * from the interval based on the function values and derivatives at its ends.
+   * @p a1 is a trial estimate of the step.
+   *
+   * The function implements Algorithms
+   * 2.6.2 and 2.6.4 on pages 34-35 in Fletcher, 2013, Practical methods of
+   * optimization. that are minor variations of the Algorithm 3.5 and 3.6 on
+   * pages 60-61 in Nocedal and Wright, Numerical optimization.
```

Comment:
I was hoping you'd mention this book -- it's one of my all-time favorite books :-)